### ISSUE TITLE
getString slow for very large ResultSets

### ISSUE THREAD

**Issue by tildedave** [xerial/sqlite-jdbc#561]
Hello, I'm having an issue with the library for large resultsets (many MB).   I can quickly write data into SQLite but (4 minutes) but reading data out takes on the order of 1-2 hours.  By "slow" it looks to be on the order of 7ms per row - unfortunately I have a few million rows.

From running jstack I've pinned it down to `column_text_utf8` and specifically the allocation of a new byte array here: https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/core/NativeDB.c#L90-L109

Example jstack (`jstack`):
```
"thread-id" #25621 prio=5 os_prio=0 tid=0x00007fa2a8028800 nid=0x38b7 runnable [0x00007fa1732b6000]
   java.lang.Thread.State: RUNNABLE
        at org.sqlite.core.NativeDB.column_text_utf8(Native Method)
        - locked <0x0000000462e008d8> (a org.sqlite.core.NativeDB)
        at org.sqlite.core.NativeDB.column_text(NativeDB.java:249)
        - locked <0x0000000462e008d8> (a org.sqlite.core.NativeDB)
        at org.sqlite.jdbc3.JDBC3ResultSet.getString(JDBC3ResultSet.java:437)
```

Mixed stack trace (`jstack -m`):
```
----------------- 14519 -----------------
0x00007fa2ce285dfa      _ZN14JNIHandleBlock15allocate_handleEP7oopDesc + 0x4a
0x00007fa2ce2584cb      jni_NewByteArray + 0x21b
```

I believe the issue here ends up being that using `getString` to pull the data out is slow, since it allocates an Java buffer here: https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/core/NativeDB.c#L99

I think this ends up being a bit of a double-memory allocation since if SQLite returns a very large dataset the driver needs to allocate a very large Java string to fit the dataset into.

Something I'm not clear on is that this doesn't happen for every query in our application (in fact, unloading some of the troublesome queries is quite fast when run at different times), so I think there must be something special about the how the driver interacts with internal memory that makes this particularly bad.  (Certain amount of native memory already allocated, insufficient pages given to the application, ...) I'm not really a JNI expert so I'm not sure what I might look at to pin this behavior down a bit better.

## Proposed Solution

I think I could override JDBCResultSet#getCharacterStream here to use a temporary (Java) buffer and pass it into SQLite, so avoiding the repeated byte buffer allocation. https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/jdbc3/JDBC3ResultSet.java#L262-L268

What this solution would look like would be something that wraps the SQLite resultset pointer and writes data from it a little at a time into a buffer allocated in Java-land.  The current implementation just uses getString under the hood so it unfortunately isn't a solution for my problem.

That said, pretty much any way I can get the data out of in-memory SQLite and into Java-land would solve my problems.  (I think even writing it to a file would let me solve this, though that comes with its own issues.)

## Conclusion

Thanks for any guidance you can provide, I will probably see if I can make my `getCharacterStream` idea work for my application and promote a patch upstream if it sounds like something that would be compatible with the library.

**Comment by xerial** [comment]
Thanks for the detailed report. The corresponding code is a very ancient one and there would be a way to avoid copying memory regions from SQLite -> String. 

- NativeDB.c:  Copying the memory region allocated by SQLite into Java's byte array (1st copy)
- JDBC3ResultSet: Converting this byte array into String  (2nd copy and UTF8 processing)

One possible option would be returning DirectByteBuffer instead of copying to byte arrays. DirectByteBuffer can read raw memory pointer. A concern is the lifetime of the allocated memory region by SQLite, but when building a String object, the memory content will be copied once, so using this String object will be safe.

**Comment by xerial** [comment]
@tildedave If you can provide the distribution of String length (e.g., p50, p95, p99, max length) in your data, it will be helpful to understand what needs to be optimized.

**Comment by xerial** [comment]
Depending on your use cases, simply reading data as `byte[]` can be a solution, instead of converting utf8 bytes into Java String (UTF-16 encoding). If that satisfies your requirement, extending getBytes implementation will improve the performance just by having a single data copy.

**Comment by tildedave** [comment]
Thanks for the comments.  I have a few new directions to go now based on them so I really appreciate it.

I implemented the `getCharacterStream` approach that I originally mentioned above and it has improved things somewhat.  To make that approach work with the Java Reader interface (takes char[]s) I needed to convert to UTF-16 on the SQLite side.  I still have some longer conversion times though the performance is overall much better.

I'm going to experiment with the DirectByteBuffer approach that you mentioned and see if I can make that work.

**Comment by tildedave** [comment]
OK, switching everything to allocate a DirectByteBuffer in JNI and pass that back up works like a charm.  I'll open a PR.

**Comment by tildedave** [comment]
Thanks much for the guidance.  I'll let you know if I find other similar-ish things that end up speeding up my larger operations.

**Comment by xerial** [comment]
Released 3.34.0 with this change. Thanks for the contribution! https://github.com/xerial/sqlite-jdbc/releases/tag/3.34.0
